netwaystev2/src/filter/filter.rs

```diff
@@ -0,0 +1,756 @@
+
+pub struct Filter {
+    transport_cmd_tx:    TransportCmdSend,
+    transport_rsp_rx:    Option<TransportRspRecv>,    // TODO no option
+    transport_notice_rx: Option<TransportNotifyRecv>, // TODO no option
+    filter_cmd_rx:       Option<FilterCmdRecv>,       // TODO no option
+    filter_rsp_tx:       FilterRspSend,
+    filter_notice_tx:    FilterNotifySend,
+    mode:                FilterMode,
+    per_endpoint:        HashMap<Endpoint, FilterEndpointData>,
+    phase_watch_tx:      Option<watch::Sender<Phase>>, // Temp. holding place. This is only Some(...) between new() and run() calls
+    phase_watch_rx:      watch::Receiver<Phase>,
+    ping_endpoints:      HashMap<Endpoint, (LatencyFilter, PingPong, Option<ProcessUniqueId>)>,
+}
+
+impl Filter {
+    pub fn new(
+        transport_cmd_tx: TransportCmdSend,
+        transport_rsp_rx: TransportRspRecv,
+        transport_notice_rx: TransportNotifyRecv,
+        mode: FilterMode,
+    ) -> FilterInit {
+        let (filter_cmd_tx, filter_cmd_rx): (FilterCmdSend, FilterCmdRecv) = mpsc::channel(FILTER_CHANNEL_LEN);
+        let (filter_rsp_tx, filter_rsp_rx): (FilterRspSend, FilterRspRecv) = mpsc::channel(FILTER_CHANNEL_LEN);
+        let (filter_notice_tx, filter_notice_rx): (FilterNotifySend, FilterNotifyRecv) =
+            mpsc::channel(FILTER_CHANNEL_LEN);
+
+        let per_endpoint = HashMap::new();
+        let ping_endpoints = HashMap::new();
+
+        let (phase_watch_tx, phase_watch_rx) = watch::channel(Phase::Running);
+
+        (
+            Filter {
+                transport_cmd_tx,
+                transport_rsp_rx: Some(transport_rsp_rx),
+                transport_notice_rx: Some(transport_notice_rx),
+                filter_cmd_rx: Some(filter_cmd_rx),
+                filter_rsp_tx,
+                filter_notice_tx,
+                mode,
+                per_endpoint,
+                phase_watch_tx: Some(phase_watch_tx),
+                phase_watch_rx,
+                ping_endpoints,
+            },
+            filter_cmd_tx,
+            filter_rsp_rx,
+            filter_notice_rx,
+        )
+    }
+
+    pub async fn run(&mut self) {
+        let transport_cmd_tx = self.transport_cmd_tx.clone();
+        let transport_rsp_rx = self.transport_rsp_rx.take().unwrap();
+        let transport_notice_rx = self.transport_notice_rx.take().unwrap();
+        let phase_watch_tx = self.phase_watch_tx.take().unwrap();
+        tokio::pin!(transport_cmd_tx);
+        tokio::pin!(transport_rsp_rx);
+        tokio::pin!(transport_notice_rx);
+
+        let filter_cmd_rx = self.filter_cmd_rx.take().unwrap();
+        let filter_rsp_tx = self.filter_rsp_tx.clone();
+        let filter_notice_tx = self.filter_notice_tx.clone();
+        tokio::pin!(filter_cmd_rx);
+        tokio::pin!(filter_rsp_tx);
+        tokio::pin!(filter_notice_tx);
+
+        let mut ping_interval_stream = tokio::time::interval(Duration::new(2, 0));
+        ping_interval_stream.tick().await;
```

Comment:
@Manghi adding this line causes the `basic_server_filter_flow` test to fail.

Reply:
Fixed.

Reply:
Nice catch with the `tokio::task::yield_now`!